Added EagerRegisterPipe

SaschaPeukert opened the pull request.

SaschaPeukert committed:
Added EagerRegisterPipe

alexaverbuch committed:
Specifies correct scenarios to test against to make EagerizationAcceptanceTest more green

alexaverbuch committed:
Adds ref column support to EagerRegisterPipe

alexaverbuch committed:
Fixes bug where registered eager would not collect reference slots
Adds setLongAt/setRefAt/longs/refs to anonymous ExecutionContext returned by registered eager

alexaverbuch commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/EagerRegisterPipe.scala:
```diff
@@ -0,0 +1,117 @@
+
+          override def setRefAt(offset: Int, value: Any): Unit = refRows(offset + globalRefOffset) = value
+
+          // if this is called from somewhere like PrimitiveExecutionContext#copyFrom the array copy is redundant
```
@systay what are your thoughts about this? also, generally about this pipe?

alexaverbuch committed:
Fixes bug in EagerRegisterPipe.ExecutionContext#longs()/refs()
 - Incorrect offsets were used

alexaverbuch committed:
Introduces ResizeableArray to EagerRegisterPipe

alexaverbuch commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/EagerRegisterPipe.scala:
```diff
@@ -98,12 +101,42 @@ case class EagerRegisterPipe(source: Pipe, pipelineInformation: PipelineInformat
+
+    protected def newArray(size: Int): Array[T]
+
+    // NOTE: current impl doesn't reuse space of old array. Instead, it creates bigger array & copies old elements to it
```
@systay @eebus if/when reviewing PR please pay special attention to `ResizableArray`.
It may turn out you want to revert this commit, to go back to using `ArrayBuffer` (FYI, all tests still pass if this commit is reverted).
The benefit of aggregating into `Array` is we get to `arraycopy` in many places, which (presumably) becomes more beneficial with more columns.
Downside of current impl is (as in comment) every time we resize array we *copy* elements to new array. It's done with `arraycopy` so hopefully low cost, but it *does* mean that we waste space (and some time). 
In worst case this impl consumes 3x more space than necessary -- if we reused old array and "appended arrays" (as in your branch, Andres) then worst case would be 2x (when we are one element past resize border). Maybe we want to make that addition to `ResizableArray`, but it would complicate reading. cc @SaschaPeukert

systay commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala:
```diff
@@ -160,6 +160,11 @@ object RegisterAllocation {
         pipeline
 
+      case Eager(source) =>
+        val pipeline = allocate(source,nullable,argument)
```
I've been thinking about this one - shouldn't this be a pipeline breaker?

SaschaPeukert committed:
Made Eager a pipeline breaker

SaschaPeukert committed:
Fixing tests

alexaverbuch committed:
Fixes failing EagerizationAcceptanceTest

alexaverbuch committed:
Refactors UpdateReportingAcceptanceTest to use CypherComparisonSupport

alexaverbuch committed:
Refactors NodeIndexSeekByRangeAcceptanceTest to use CypherComparisonSupport

alexaverbuch committed:
Refactor EagerRegisterPipe to use AnyValue

alexaverbuch committed:
Rewrites registered eager pipe to be greatly simplified
 - Previous impl was not written with pipeline breaking in mind
 Corrected eager test for pipeline breaking in EnterprisePipeBuilderTest

alexaverbuch committed:
Removes acceptance test from cost-interpreted-enterprise blacklist
 - "Combine MATCH, WITH and CREATE"

alexaverbuch committed:
Run more MatchAcceptanceTest tests with Slotted scenario
Fix error where CompatibilitySpecSuiteTest.SCENARIO_NAME_REQUIRED was left non-empty

alexaverbuch committed:
Adds "override" & type annotations to methods in ExecutionContext impls

alexaverbuch committed:
Fixes compilations errors in tests due to Enterprise --> Slotted rename

alexaverbuch committed:
Fixes bug in EagerRegisterPipe, actually makes it eager

alexaverbuch committed:
Expect acceptance test for node index seek to pass against slotted runtime

alexaverbuch committed:
Makes EagerizationAcceptanceTest green

alexaverbuch committed:
Runs EagerizationAcceptanceTest tests against more scenarios

systay commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/ExecutionContext.scala:
```diff
@@ -78,32 +85,32 @@ case class MapExecutionContext(m: MutableMap[String, AnyValue])
   }
 
-  override def +=(kv: (String, AnyValue)) = {
+  override def +=(kv: (String, AnyValue)): MapExecutionContext.this.type = {
```
These types are IntelliJ artifacts and not helpful/pretty. In this case, you should just use `MapExecutionContext`.

systay replied:
Never mind me. ^ That was a lie.

systay commented:
Really nice that you have taken on updating `CypherComparisonSupport` like this. 👍
